client/quantum_serverless/core/provider.py

```diff
@@ -147,7 +147,27 @@ def __repr__(self):
 
 
 class BaseProvider(JsonSerializable):
-    """Provider."""
+    """
+    The base class for Quantum Serverless providers.
+
+    Example:
+        >>> provider = ServerlessProvider(
```

Comment:
nit
```suggestion
        >>> provider = BaseProvider(
```

Reply:
also, since these args are for the init method, should we leave that docstring there? (I'm not 100% on the python best practices here, so not sure what the answer is)

Reply:
Ya, I'm conflicted here, since this code won't run. Maybe I just move this example to ``ServerlessProvider``. I could duplicate it and remove ``host`` for ``IBMServerlessProvider``. It's a little redundant, but more correct than this.

Reply:
Regarding the args, I took a quick glance at `qiskit.QuantumCircuit`, and they have their args there. It's probably better to just keep things consistent with the rest of the serverless repo. I can move the examples and docstrings back inside the init methods.

Reply:
> Maybe I just move this example to ServerlessProvider

I think that makes sense, especially as it'll show on the site in the API reference: https://qiskit-extensions.github.io/quantum-serverless/stubs/quantum_serverless.core.BaseProvider.html

though, looking at that link, it looks like the current version has the same issue...

Reply:
maybe we update the docstring for `BaseProvider` to somehow note that it isn't implemented (?)

Reply:
@psschwei , I've turned `BaseProvider` into a proper abstract base class and flagged the class methods which should be implemented by subclasses. I've moved args down into constructors with a generic constructor description and left the examples and class descriptions up at the class level. Let me know if this looks good to you

Reply:
@IceKhan13 , any reason `BaseProvider` is used directly all throughout the [quantum_serverless](https://github.com/Qiskit-Extensions/quantum-serverless/blob/main/client/quantum_serverless/quantum_serverless.py) module? I feel like this should be an abstract base class, and that `ServerlessProvider` should be the most general, usable class.

This would be a breaking change, since you're already using this within `quantum_serverless`, so I hesitate to do it. I could either raise a deprecation warning inside the `BaseProvider` constructor and we could make it an abstract class later, or I could do a breaking change now to make it abstract. I only suggest the breaking change because this class isn't featured in any of our docs, and it isn't clear to me you ever intended for users to instantiate this class. It seems `ServerlessProvider`/`Provider` give you everything you need.

Reply:
Ahh, I can't make it abstract. It has some extra functionality for specifying a compute resource that `ServerlessProvider` doesn't provide. Standby...

Reply:
`ServerlessProvider/Provider` is to connect to gateway service. 
If somebody will want to implement Azure or AWS provider, they can do that by extending base provider. 

`BaseProvider` is only used for type hints and as base class for `Provider`.

Code is using base provider in some shared cases, I think. But I agree it might need some love. I would argue though it is out of the scope of renaming pr :)

Reply:
this PR should be rename/reprecate. For refactor we will need separate PR which will need more test time

Reply:
or, keep this refactor PR and open `just rename PR` :)

Reply:
Yeah, we will do it in separate PR abstract and that is it :)